### PR TITLE
Use relative paths in tests

### PR DESCRIPTION
**Pull request by Manishearth**
in non-cargo build systems it's not necessary that the build time environment will be identical to the run time environment. This ensure that the variable is computed at the time it is actually accessed.

### CODE REVIEW

File: provider/datagen/src/provider.rs
Comment (robertbastian) [inline provider/datagen/src/provider.rs:72]: I think a good default would be to `unwrap_or_default()` if the variable isn't set, which will resolve to the current directory. If we want to be more compatible with non-cargo, we shouldn't force users to set that flag, instead they could run the tests in the appropriate location.

File: provider/datagen/src/lib.rs
Comment (robertbastian) [inline provider/datagen/src/lib.rs:693]: Given that this is a single file, I'd prefer to `include_bytes!` it and write it to a temp file in order to call the API. This way it's not required to be there at runtime.

File: provider/datagen/src/provider.rs
Comment (Manishearth) [reply provider/datagen/src/provider.rs:72]: That seems more prone to bugs IMO. I'd rather make that change when someone asks us to.

File: provider/datagen/src/lib.rs
Comment (Manishearth) [reply provider/datagen/src/lib.rs:693]: I don't want to muck with tempfiles in a test when it's already in a file. I'm fine with requiring it at runtime.

Comment (robertbastian) [reply provider/datagen/src/lib.rs:693]: I think to test a method that accepts `Path` we should use a mock file system. Rust's mock file system is `/tmp`.

In the bigger picture I want to get to a point where any test that uses a single file just `includes!` it. `CARGO_MANIFEST_DIR` and panicking on missing files is not something I want in our tests.

File: provider/datagen/src/provider.rs
Comment (robertbastian) [reply provider/datagen/src/provider.rs:72]: What kind of bugs? I think requiring `CARGO_MANIFEST_DIR` is overly restricting. It probably works for us, but I don't see a need for this panic.

**Review by robertbastian — COMMENTED** [review]
I'm really not a fan of any of these changes.

> in non-cargo build systems it's not necessary that the build time environment will be identical to the run time environment.

Well, we use cargo, maybe the non-cargo build systems have to bridge this gap somehow.

File: provider/datagen/src/lib.rs
Comment (Manishearth) [reply provider/datagen/src/lib.rs:693]: > Rust's mock file system is /tmp

This would just break on Windows.

File: provider/datagen/src/provider.rs
Comment (Manishearth) [reply provider/datagen/src/provider.rs:72]: Because this now depends on where you run `cargo test` from, and if you get that wrong you'll get a file not found error that will be annoying to debug.

**Comment by Manishearth** [comment]
> Well, we use cargo, maybe the non-cargo build systems have to bridge this gap somehow.

And multiple of our users don't. We should try and make concessions where possible, I don't see us as losing much here (the code does not change behavior for ICU4X devs), and it's slightly nicer for others.

File: provider/datagen/src/lib.rs
Comment (robertbastian) [reply provider/datagen/src/lib.rs:693]: I mean `std::env::temp_dir` obviously. Please try to address my actual concern.

**Comment by robertbastian** [comment]
> multiple of our users don't

Who? Doing this does not completely solve the problem for Google, as it will still require manual setup for each test target. We're currently internally discussing a solution that would not need this PR, so let's maybe wrap that up first?

**Review by sffc — COMMENTED** [review]
Not super excited that this makes tests not runnable outside cargo (e.g. if you run the test binary directly without `cargo test`).

How about if we check the runtime env, then fall back to the build time env, and only then do we fail?

File: provider/datagen/src/lib.rs
Comment (Manishearth) [reply provider/datagen/src/lib.rs:693]: I didn't know that existed! I'm still iffy on that; since there can be other stuff in the tempdir, and non-Cargo build systems may also not like files getting written to `/tmp` (many such build systems are hermetic for reproducability)

**Comment by Manishearth** [comment]
> Who?

gn was having similar problems with a bunch of build time stuff.

Distros also do care about tests and don't invoke `cargo`, while we haven't interacted with them much as ICU4X I'm reminded of similar 


> We're currently internally discussing a solution that would not need this PR,  so let's maybe wrap that up first?

I'd love for things to be nicer there but it's not the first time I've had a conversation about this, it's not a new idea being presented to them. The details escape me and things may have changed but I'm generally pessimistic as to a complete fix from the Bazel side. I don't think a full solution has been proposed yet that addresses all of the concerns.

for context for non-Googlers: many general purpose build systems, including Bazel and gn, don't necessarily run the build and run steps from the same folder, environment, or even machine. If the path passed in to CARGO_MANIFEST_DIR changes, then the test needs to read the test-time path, not the build time path. Potential fixes include always passing in relative paths at build and run time, though there are some tricky bits there as to how that would interact with things like `include!()`.

> Not super excited that this makes tests not runnable outside cargo

Yeah, and I guess that impacts `gdb`-ability. Hm.

> How about if we check the runtime env, then fall back to the build time env, and only then do we fail?

wfm.

File: provider/datagen/src/lib.rs
Comment (Manishearth) [reply provider/datagen/src/lib.rs:693]: "writes to a file during a test" is the kind of thing that causes flakes in my experience

Comment (Manishearth) [reply provider/datagen/src/lib.rs:694]: If the goal is to use more `include!`s (which I'm very much in favor of) I think the better option is to refactor `keys_from_file()` and `keys_from_bin()` into a pair of functions, one of which just takes a bag of bytes.

**Comment by Manishearth** [comment]
I'd really like this to make 1.4

File: provider/datagen/src/lib.rs
Comment (robertbastian) [reply provider/datagen/src/lib.rs:694]: please do that then

File: provider/datagen/src/lib.rs
Comment (robertbastian) [inline provider/datagen/src/lib.rs:676]: the types don't work out here. Once you add the conversions this expression will be complex enough to factor it out

File: provider/datagen/src/lib.rs
Comment (sffc) [inline provider/datagen/src/lib.rs:678]: Nit: You changed some of the other examples to use `Path::join`; please be consistent and use that here as well. Note that `keys_from_file` takes an `AsRef<Path>`.

**Review by sffc — APPROVED** [review]

File: provider/datagen/src/lib.rs
Comment (Manishearth) [reply provider/datagen/src/lib.rs]: yeah I was waiting on a final call on what we're supposed to do here before fixing everything

**Comment by Manishearth** [comment]
> This seems like an incomplete solution. We should release this behaviour across all crates or none.

sure, we only do this in two other (non-experimental, non-internal, non-deprecated) places: fs data provider and gen-greek-to-me.

**Review by sffc — DISMISSED** [review]

**Comment by robertbastian** [comment]
I think datetime as well? Anywhere we use CARGO_MANIFEST_DIR

**Comment by Manishearth** [comment]
I don't see it.

**Comment by Manishearth** [comment]
datetime just uses relative paths

**Review by sffc — DISMISSED** [review]
CI is still failing

**Comment by robertbastian** [comment]
ah datetime uses includes, which I don't think is brittle because it's compile time resolved?

File: components/casemap/tests/gen_greek_to_me.rs
Comment (robertbastian) [inline components/casemap/tests/gen_greek_to_me.rs:171]: `include` is always resolved relative to the source file, so there's no need to involve `CARGO_MANIFEST_DIR`

also everywhere below

**Comment by Manishearth** [comment]
Not the fixture tests, but yes, it uses includes

File: provider/datagen/src/provider.rs
Comment (robertbastian) [inline provider/datagen/src/provider.rs:71]: I really don't like the complexity of this. 6 lines to define a path is 5 too many.

**Comment by robertbastian** [comment]
> sure, we only do this in two other (non-experimental, non-internal, non-deprecated) places: fs data provider and gen-greek-to-me.

Why do you consider experimental out of scope? I get 28 results for `CARGO_MANIFEST_DIR` at head, this fixes 8. `icu_provider_fs` docs tests for example use it, those are stable tests that should be treated the same as unit tests.

File: provider/fs/tests/test_file_io.rs
Comment (robertbastian) [inline provider/fs/tests/test_file_io.rs:18]: naming suggestion:

```suggestion
    let maybe_root = std::env::var("CARGO_MANIFEST_DIR").ok();
    let root = maybe_env
        .as_deref()
        .or(option_env!("CARGO_MANIFEST_DIR"))
        .expect("icu_provider_fs tests need CARGO_MANIFEST_DIR to run");
    let root = PathBuf::from(dir);
```

**Review by sffc — DISMISSED** [review]

**Comment by Manishearth** [comment]
> Why do you consider experimental out of scope? I get 28 results for `CARGO_MANIFEST_DIR` at head, this fixes 8.

Because I don't have time to do the whole thing right now in time for the release, and I don't want to delay the release on this, but I would like to get this in time for datagen in release.

I don't agree with the "all or nothing" stance in the first place; I'm happy to fix a couple other places but I think incremental improvements are fine here.

Ultimately, "ICU4X in non-Cargo build systems" is going to always be based on the problems non-Cargo build systems report to us; we're not going to be perfect, we can just try piece by piece as we notice things. I'm fine with this being a partial fix. I don't agree with needing to be perfect here.

**Comment by Manishearth** [comment]
Also most of those 28 are doctests. I don't think we should fix those; that would make our docs uglier.

**Comment by Manishearth** [comment]
I changed my mind on relative paths; since we already use them elsewhere and while the errors may be confusing it'll still be predictable and usually easy to fix. I wasn't a fan of the verbosity of the unwrap solution either.

I've also fixed transliterator and resb.

I do not plan to fix the doctests, internal tools, or testdata. Doctests are anyway an extremely cargo+rustdoc concept, they're hard to run outside of cargo and basically nobody does (to the extent that even _miri_ only recently got support). Internal tools should only be run by us, so can use whatever they want. Testdata is deprecated.


I do think this PR has already expanded in scope quite a bit and I'd really like this to land for 1.4 since it is still an incremental improvement.

**Comment by robertbastian** [comment]
> I changed my mind on relative paths; since we already use them elsewhere and while the errors may be confusing it'll still be predictable and usually easy to fix. I wasn't a fan of the verbosity of the unwrap solution either.

I've looked into this a bit more: `cargo test` executes the test binary inside the crate (verified with `std::env::current_dir()`), so relative paths will work with cargo. Other build systems that want to execute cargo tests should reproduce this behaviour, which I don't think is very hard.

This also applies to `cargo test --doc`, so we can also remove `CARGO_MANIFEST_DIR` from all docs test. Even if you don't want to run them outside cargo, this makes the docs more readable. While we're at this I'd like to get this into 1.4 as a docs improvement; I can do this if you don't want to.

Note that `cargo run` does not do this, so scripts are the only place where we need to keep the `CARGO_MANIFEST_DIR`s. 

I see you're still doing the verbose thing in `provider/datagen/src/provider.rs`, did you mean to change that?

**Comment by Manishearth** [comment]
Yeah, provider.rs was an oversight. I'm off for today and tomorrow is SAH so this is probably as far as this PR is going to get; would be great if you could do the other cleanups.

**Comment by robertbastian** [comment]
@sffc please review my changes

**Comment by robertbastian** [comment]
@Manishearth if you approve in a comment I'll merge

**Comment by Manishearth** [comment]
oh yeah I saw your commits, r+

**Review by robertbastian — APPROVED** [review]

**Comment by sffc** [comment]
I'm confused. I approved this PR but the PR like completely changed since I approved it, in ways we didn't discuss. I will take a look but now this is landed.

File: experimental/transliterate/tests/cldr.rs
Comment (sffc) [inline experimental/transliterate/tests/cldr.rs:15]: Issue: This requires that the test is run in a certain directory, right? That's the whole point of using `CARGO_MANIFEST_DIR`.

File: provider/adapters/src/fork/predicates.rs
Comment (sffc) [inline provider/adapters/src/fork/predicates.rs:80]: I care less about docs tests since docs tests are pretty much only ever run in `cargo`. I'm more concerned about integration tests since those are standalone binaries.

**Comment by sffc** [comment]
> I've looked into this a bit more: cargo test executes the test binary inside the crate (verified with std::env::current_dir()), so relative paths will work with cargo. Other build systems that want to execute cargo tests should reproduce this behaviour, which I don't think is very hard.

I just really don't agree with this sentiment. A test binary is a test binary and I shouldn't need to `cd` to a particular directory in order to run it. Test binaries are useful in external contexts such as being run with a fuzzer, debugger, profiler, etc. This is an unnecessary headache to impose when we have a clear, clean way of avoiding it.

I really like the "first try runtime environment path and the fall back to compile time environment path" approach. It is flexible but still easy to use.

> I changed my mind on relative paths; since we already use them elsewhere and while the errors may be confusing it'll still be predictable and usually easy to fix.

Where do we use relative paths? We should fix them up to use the runtime-or-build-time-environment approach. In other words, we're moving in directly the wrong direction.

> I wasn't a fan of the verbosity of the unwrap solution either.

It really is not that verbose. If we care about the verbosity we can make a helper macro or something.

**Comment by sffc** [comment]
Just as an example, cargo works:

```bash
$ cargo test -p icu_provider_fs --all-features --test test_file_io
     Running tests/test_file_io.rs (target/debug/deps/test_file_io-6ab31c1702a4f7e7)

running 2 tests
test test_errors ... ok
test test_provider ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

but then the binary doesn't work by itself:

```bash
$ target/debug/deps/test_file_io-6ab31c1702a4f7e7

running 2 tests
test test_errors ... FAILED
test test_provider ... FAILED

failures:

---- test_errors stdout ----
thread 'test_errors' panicked at provider/fs/tests/test_file_io.rs:55:54:
called `Result::unwrap()` on an `Err` value: DataError { kind: Io(NotFound), key: None, str_context: None, silent: false }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- test_provider stdout ----
thread 'test_provider' panicked at provider/fs/tests/test_file_io.rs:20:54:
called `Result::unwrap()` on an `Err` value: DataError { kind: Io(NotFound), key: None, str_context: None, silent: false }


failures:
    test_errors
    test_provider

test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
```

so now to make it work I need to do things like this:

```bash
$ (cd provider/fs && ../../target/debug/deps/test_file_io-6ab31c1702a4f7e7)

running 2 tests
test test_errors ... ok
test test_provider ... ok

test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.01s
```

That's definitely a regression, not an improvement.

**Comment by sffc** [comment]
Why do I care about this? Because I want to run commands like

```
$ valgrind target/debug/deps/test_file_io-6ab31c1702a4f7e7
$ gdb target/debug/deps/test_file_io-6ab31c1702a4f7e7
```

**Comment by robertbastian** [comment]
Well, it's pretty simple, the binary is meant to run inside the crate. If you move the crate after compiling, or it's run on a different machine, the binary compiled with `CARGO_MANIFEST_DIR` will also not work, because it hardcodes a path on the compiling machine. That is worse, because it's much easier to run it in a copy of the crate than to run it with a copy of the crate at the exact same absolute path.

**Comment by sffc** [comment]
That's why I endorsed the solution that allows a runtime-provided `CARGO_MANIFEST_PATH`. My use cases up to this point didn't involve running on different machines than the code was compiled on, and that's a totally valid use case, but we should tack it on without regressing the use case of running the test binaries directly from `target` on the same machine.

**Comment by robertbastian** [comment]
```
$ (cd provider/fs && valgrind ../../target/debug/deps/test_file_io-6ab31c1702a4f7e7)
$ (cd provider/fs && gdb ../../target/debug/deps/test_file_io-6ab31c1702a4f7e7)
```

**Comment by sffc** [comment]
Also, this adversely impacts fixture tests more than other types of tests. We generally don't want to put fixture tests at a disadvantage. Reading data from a test file is always a pain; `CARGO_MANIFEST_DIR` makes it substantially less of a pain. The _only_ side effect I'm aware of is this one where you want to run tests on a different machine, and so we have a good workaround for that, by providing the environment variable at runtime.

Yes there are ways to run things with relative paths. My position is pretty clear and consistent with my position on most issues: we have several constituencies of the following order of priority:

1. End users
2. Clients of ICU4X
3. ICU4X contributors
4. ICU4X core team members

What's better for end users should take priority over what's better for clients, and what's better for clients should take priority over what's better for contributors, and what's better for contributors should take priority over what's better for core team members.

My position is that regressing to requiring relative paths to run integration test binaries adversely impacts constituencies 2 and 3. On the other hand, the 5-like environment variable boilerplate adversely impacts constituencies 3 and 4. Therefore, the environment variable boilerplate is the better solution.

**Comment by Manishearth** [comment]
> Also, this adversely impacts fixture tests more than other types of tests.

Our fixture tests have been using relative paths for _years_ though

**Comment by Manishearth** [comment]
Also I don't see how the relative paths thing affects anyone other than ICU4X devs.

**Comment by sffc** [comment]
Clients of ICU4X want to run tests; if they didn't care about running tests then we never would have needed to make this change in the first place

I'm aware that we had some tests that already used relative paths. We were inconsistent in the repo. In the sentence "this adversely impacts fixture tests" the "this" refers to relative path file includes, not to the PR in question, which doesn't touch fixture tests.

**Comment by Manishearth** [comment]
Yeah but currently these tests are broken due to things out of clients' control (assuming they do not control their choice of build system). The assumption being made here is that the build and test environments are the same, an assumption that does not pan out for most general purpose build systems.

This changes it to requiring certain files to be available in certain paths, which most build systems do let you control. I agree environment variables give you more control than pwd and don't mind going back to the other solution if Rob can be convinced.

But so far "we have to run tests from the repo-not-crate root" is not a build system feature I have seen often. What I have seen is that it is more convenient for humans doing this manually like your valgrind example, and that seems to fall into user base 4.


(Not saying we should rerevert this for 1.4, to be clear, just putting down thoughts for when we do have this discussion)

**Comment by sffc** [comment]
I think I agree with everything above. I only disagree on using relative paths instead of the runtime environment override.

**Pull request merged**
